### ISSUE TITLE
GPUParticles3D GPU artifacts (only in exported build, works fine in DEBUG mode)

### ISSUE THREAD

**Issue by curvednebula** [godotengine/godot#82185]
### Godot version

v4.1.1.stable.mono.official [bd6af8e0e]

### System information

MacOS 13.5.2 (Macbook M1 Pro, 16Gb) - Mobile and Forward+

### Issue description

I've created a minimal project that reproduces the issue: [https://github.com/curvednebula/godot-particle-issue](https://github.com/curvednebula/godot-particle-issue)

Notes:
- The GPU artifacts issue happens only when running exported binary (Macbook M1 Pro). No issue when running from the editor.
- There is a disabled MultiMeshFlock - the problem disappears if you remove it.
- There is gradient-based coloring of the particles in Process Material - if you disable it - the problem disappears.

### Steps to reproduce

1. download the test project [https://github.com/curvednebula/godot-particle-issue](https://github.com/curvednebula/godot-particle-issue)
2. Export it to a binary (MacOS M1)
3. Run and wait for a few seconds - GPU artifacts will appear shortly.

### Minimal reproduction project

[https://github.com/curvednebula/godot-particle-issue](https://github.com/curvednebula/godot-particle-issue)

**Comment by bitsawer** [comment]
Have you tried with the latest 4.2 dev5? Or with Windows or Linux in case it also affects other than Mac systems?

The bug looks strangely specific if it is affected by seemingly unrelated things like MultiMeshFlock node presence and particle mesh type (mentioned in the Reddit thread). I gave the project a a few test run on Windows 10 with a Godot master branch ASAN editor build (but not export) just in case there might be some memory corruption involved, but at least on my system it came out clean.

**Comment by curvednebula** [comment]
No, I didn't try on Windows - currently my only dev machine is Macbook. I will definitely try on 4.2 release when it is out.

**Comment by curvednebula** [comment]
I've been thinking about this issue, and I think it looks as if variable overflow or precision issue. I think we need to check if shaders are compiled differently in debug mode and release modes?

**Comment by Calinou** [comment]
> Are shaders compiled differently in debug mode and release modes?

Not that I know of.
